{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5dcf5a39-b225-44e0-adb8-a54320652aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import cv2\n",
    "import numpy as np\n",
    "from numpy import linalg as LA\n",
    "import pandas as pd\n",
    "import math\n",
    "import traceback"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "580498ce-9b4c-46fb-8eff-bdcefc3b386d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 中心座標を求める関数\n",
    "def centerCoordinate(coordinates):\n",
    "    value = 0\n",
    "    for coor in coordinates:\n",
    "        value += coordinates[coor]\n",
    "    return value / 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "894c64a8-6e66-4815-8d3b-2978069102d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 物体検出に必要な対応点の数の下限\n",
    "MIN_MATCH_COUNT = 5\n",
    "\n",
    "# テンプレート画像の読み込み\n",
    "template = cv2.imread(\"/Users/yuki-f/Documents/SocSEL/Research/ImageRecognition/cat.png\")\n",
    "\n",
    "# 検索対象作品のスクリーンショットがあるディレクトリを指定\n",
    "path = \"/Users/yuki-f/Documents/SocSEL/Research/Selenium/dataset/sample/screenshots\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "2a0bae6b-9b65-477b-9f77-af6ceee4fd33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0      1      2\n",
      "0  0  174.5  310.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  1  262.5  310.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  2  334.5  310.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  3  406.5  310.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  4  430.5  262.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  5  430.5  190.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  6  430.5  110.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1     2\n",
      "0  7  406.5  62.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1     2\n",
      "0  8  334.5  62.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1     2\n",
      "0  9  262.5  62.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  10  182.5  62.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  11  110.5  62.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1     2\n",
      "0  12  54.5  78.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  13  54.5  150.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  14  54.5  230.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  15  54.5  302.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1      2\n",
      "0  16  118.5  262.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1      2\n",
      "0  17  190.5  208.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1      2\n",
      "0  18  262.5  154.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  19  342.5  94.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  20  366.5  58.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  21  294.5  58.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  22  222.5  58.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0      1     2\n",
      "0  23  150.5  58.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1     2\n",
      "0  24  70.5  58.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  25  54.5  114.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  26  54.5  186.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  27  54.5  258.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "    0     1      2\n",
      "0  28  54.5  306.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0     1      2\n",
      "0  0  82.5  289.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  1  122.5  259.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  2  158.5  232.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  3  198.5  202.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  4  234.5  175.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  5  270.5  148.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1      2\n",
      "0  6  306.5  121.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1     2\n",
      "0  7  346.5  91.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n",
      "   0      1     2\n",
      "0  8  382.5  64.0\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-89-22d26588566a>\", line 76, in <module>\n",
      "    df = df.append([df, addRow], axis=1)\n",
      "TypeError: append() got an unexpected keyword argument 'axis'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for pathName, dirName, fileNames in os.walk(path):\n",
    "    fileLen = len([fileName for fileName in fileNames if not fileName.startswith(\".\")])\n",
    "    \n",
    "    # ディレクトリが空（スクショが存在しない）の場合はスキップ\n",
    "    if fileLen == 0:\n",
    "        continue\n",
    "        \n",
    "    prjId = pathName.rsplit(\"/\")[len(pathName.rsplit(\"/\")) - 1]\n",
    "    \n",
    "    coordinates = []\n",
    "    \n",
    "    # csvに出力するためのdataframeを作成\n",
    "    df = pd.DataFrame(columns=[\"time\", \"x\", \"y\"])\n",
    "    \n",
    "    for i in range(fileLen):\n",
    "        fileName = pathName + \"/\" + prjId + \"-\" + str(i) + \".png\"\n",
    "        screenshot = cv2.imread(fileName)\n",
    "        \n",
    "        # 検出器生成\n",
    "        sift = cv2.xfeatures2d.SIFT_create()\n",
    "        \n",
    "        # 画像の特徴量（kp:特徴点の座標情報等, des:特徴量記述子）\n",
    "        try:\n",
    "            kp_t,des_t = sift.detectAndCompute(template,None)\n",
    "            kp_s,des_s = sift.detectAndCompute(screenshot,None)\n",
    "\n",
    "            FLANN_INDEX_KDTREE = 0\n",
    "            index_params = dict(algorithm=FLANN_INDEX_KDTREE,tress=5)\n",
    "            search_params = dict(checks = 50)\n",
    "\n",
    "            flann = cv2.FlannBasedMatcher(index_params, search_params)\n",
    "            matches = flann.knnMatch(des_t,des_s,k=2)\n",
    "\n",
    "            good = []\n",
    "            for m,n in matches:\n",
    "                if m.distance < 0.7 * n.distance:\n",
    "                    good.append(m)\n",
    "\n",
    "            # 対応点がMIN_MATCH_COUNT以上あれば物体を検出\n",
    "            if len(good) > MIN_MATCH_COUNT:\n",
    "                src_pts = np.float32([ kp_t[m.queryIdx].pt for m in good ]).reshape(-1,1,2)\n",
    "                dst_pts = np.float32([ kp_s[m.trainIdx].pt for m in good ]).reshape(-1,1,2)\n",
    "\n",
    "                M, mask = cv2.findHomography(src_pts, dst_pts, cv2.RANSAC,5.0)\n",
    "                matchesMask = mask.ravel().tolist()\n",
    "\n",
    "                h,w = template.shape[:2]\n",
    "                pts = np.float32([ [0,0],[0,h-1],[w-1,h-1],[w-1,0] ]).reshape(-1,1,2)\n",
    "                dst = cv2.perspectiveTransform(pts,M)\n",
    "\n",
    "                screenshot = cv2.polylines(screenshot,[np.int32(dst)],True,255,3, cv2.LINE_AA)\n",
    "                \n",
    "                flattenCoors = np.int32(dst).flatten()\n",
    "                \n",
    "                # 検出した物体の四隅の座標\n",
    "                coordinates.append(\n",
    "                    {\n",
    "                        \"index\": i,\n",
    "                        \"x\": {\n",
    "                            \"upperLeft\": flattenCoors[0], \n",
    "                            \"lowerLeft\": flattenCoors[2],\n",
    "                            \"lowerRight\": flattenCoors[4],\n",
    "                            \"upperRight\": flattenCoors[6]\n",
    "                        },\n",
    "                        \"y\": {\n",
    "                            \"upperLeft\": flattenCoors[1], \n",
    "                            \"lowerLeft\": flattenCoors[3],\n",
    "                            \"lowerRight\": flattenCoors[5],\n",
    "                            \"upperRight\": flattenCoors[7]\n",
    "                        }\n",
    "                    }\n",
    "                )\n",
    "                \n",
    "                addRow = pd.DataFrame([[i, centerCoordinate(coordinates[i][\"x\"]), centerCoordinate(coordinates[i][\"y\"])]])\n",
    "                print(addRow)\n",
    "                df = df.append([df, addRow], axis=1)\n",
    "                print(df)\n",
    "                \n",
    "            else:\n",
    "                print(\"マッチングなし\")\n",
    "                matchesMask = None\n",
    "                            \n",
    "        except Exception as e:\n",
    "            print(traceback.format_exc())\n",
    "            continue \n",
    "\n",
    "    if len(coordinates) <= 0:\n",
    "        continue\n",
    "    \n",
    "    df.to_csv(prjId + \".csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c268ec-cfef-46ee-a3e6-e4307ee263ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
